[RFC] packages (vim-patch:7.4.{1384,1388,1396,1478,1479,1480,1486,1492,1499,1528,1550,1551,1552,1553,1554,1596,1649,1712,1840,1973,1986})

jamessan opened the pull request:
- [x] **vim-patch:7.4.1384**
  Problem:    It is not easy to use a set of plugins and their dependencies.
  Solution:   Add packages, ":loadopt", 'packpath'.
  
  https://github.com/vim/vim/commit/f6fee0e2d4341c0c2f5339c1268e5877fafd07cf
- [x] **vim-patch:7.4.1388**
  Problem:    Compiler warning. (Cesar Romani)
  Solution:   Initialize variable.
  
  https://github.com/vim/vim/commit/bdcd75275002c3b74015bb9bc0a01b13bb6107d4
- [x] **vim-patch:7.4.1396**
  Problem:    Compiler warnings for conversions.
  Solution:   Add type cast.
  
  https://github.com/vim/vim/commit/1daae446e58fd90f98c51ff3af8f54bfa5197751
- [x] **vim-patch:f391327**
  Updated runtime files.
  
  https://github.com/vim/vim/commit/f391327adbbffb11180cf6038a92af1ed144e907
- [x] **vim-patch:dae8d21**
  Updated runtime files
  
  https://github.com/vim/vim/commit/dae8d21dd291df6a6679a00be64e18bca0156576
- [x] **vim-patch:7.4.1478**
  Problem:    ":loadplugin" doesn't take care of ftdetect files.
  Solution:   Also load ftdetect scripts when appropriate.
  
  https://github.com/vim/vim/commit/1bdd42627d619258d0e847f217cfc1c2795f1ac5
- [x] **vim-patch:7.4.1479**
  Problem:    No testfor ":loadplugin".
  Solution:   Add a test.  Fix how option is being set.
  
  https://github.com/vim/vim/commit/863c1a9079fa340d663ccafb011729a29186d73e
- [x] **vim-patch:7.4.1480**
  Problem:    Cannot add a pack direcory without loading a plugin.
  Solution:   Add the :packadd command.
  
  https://github.com/vim/vim/commit/91715873d19a1859c08eeded7848113596e2f2bd
- [x] **vim-patch:7.4.1486**
  Problem:    ":loadplugin" is not optimal, some people find it confusing.
  Solution:   Only use ":packadd" with an optional "!".
  
  https://github.com/vim/vim/commit/f3654827368e6204608036353a0360e9e7c21e02
- [x] **vim-patch:328da0d**
  Update runtime files.
  
  https://github.com/vim/vim/commit/328da0dcb7be34b594725eef6dc98d3ea6516d69
- [x] **vim-patch:7.4.1492**
  Problem:    No command line completion for ":packadd".
  Solution:   Implement completion. (Hirohito Higashi)
  
  https://github.com/vim/vim/commit/35ca0e7a1cb6e6daef8e0052a8437801226cef19
- [x] **vim-patch:7.4.1499**
  Problem:    No error message when :packadd does not find anything.
  Solution:   Add an error message. (Hirohito Higashi)
  
  https://github.com/vim/vim/commit/be82c254862e475a582c0717455e1db6bf96b0d0
- [x] **vim-patch:5f148ec**
  Update runtime files.
  
  https://github.com/vim/vim/commit/5f148ec0b5a6cedd9129b3abac351034b83cc4f7
- [x] **vim-patch:7.4.1528**
  Problem:    Using "ever" for packages is confusing.
  Solution:   Use "start", as it's related to startup.
  
  https://github.com/vim/vim/commit/af1a0e371e739f8dff337fd31da0ff8ffb347b43
- [x] **vim-patch:77cdfd1**
  Updated runtime files.
  
  https://github.com/vim/vim/commit/77cdfd10382e01cc51f4ba1a9177032351843151
- [x] **vim-patch:7.4.1550**
  Problem:    Cannot load packages early.
  Solution:   Add the ":packloadall" command.
  
  https://github.com/vim/vim/commit/2d8f56acb32428d0f965d42dd13b27100b46fa15
- [x] **vim-patch:7.4.1551**
  Problem:    Cannot generate help tags in all doc directories.
  Solution:   Make ":helptags ALL" work.
  
  https://github.com/vim/vim/commit/6bef5306e4f2cacb3a93667992c2312d4b293c9d
- [x] **vim-patch:7.4.1552**
  Problem:    ":colorscheme" does not use 'packpath'.
  Solution:   Also use in "start" and "opt" directories in 'packpath'.
  
  https://github.com/vim/vim/commit/7f8989dd8a627af2185df381195351a913f3777f
- [x] **vim-patch:7.4.1553**
  Problem:    ":runtime" does not use 'packpath'.
  Solution:   Add "what" argument.
  
  https://github.com/vim/vim/commit/8dcf259d904cfb965d31841dc74a5cfaf5a351d9
- [x] **vim-patch:7.4.1554**
  Problem:    Completion for :colorscheme does not use 'packpath'.
  Solution:   Make it work, add a test. (Hirohito Higashi)
  
  https://github.com/vim/vim/commit/52f9c19015df5ee1ee8592b6f3f15b8a57c8f5be
- [x] **vim-patch:7.4.1596**
  Problem:    Memory leak. (Coverity)
  Solution:   Free the pattern.
  
  https://github.com/vim/vim/commit/ba8cd122ef60a7c71a7723be0d635f0c2d4556ab
- [x] **vim-patch:e18c0b3**
  Updated runtime files.
  
  https://github.com/vim/vim/commit/e18c0b39815c5a746887a509c2cd9f11fadaba07
- [x] **vim-patch:4f3f668**
  Updated runtime files.
  
  https://github.com/vim/vim/commit/4f3f668c8486444e53163c29d2fc79bf47eb3c82
- [x] **vim-patch:7.4.1649**
  Problem:    The matchit plugin needs to be copied to be used.
  Solution:   Put the matchit plugin in an optional package.
  
  https://github.com/vim/vim/commit/aedfcbe1e6c7df6edcd6756d7601bfdec7dd2087
- [x] **vim-patch:7db8f6f**
  Updated runtime files.
  
  https://github.com/vim/vim/commit/7db8f6f4f85e5d0526d23107b2a5e2334dc23354
- [x] **vim-patch:7.4.1712**
  Problem:    For plugins in packages, plugin authors need to take care of all
              dependencies.
  Solution:   When loading "start" packages and for :packloadall, first add all
              directories to 'runtimepath' before sourcing plugins.
  
  https://github.com/vim/vim/commit/49b27326447d0827c59c6cd201d58f65c1163086
- [x] **vim-patch:03413f4**
  Updated runtime files.
  
  https://github.com/vim/vim/commit/03413f44167c4b5cd0012def9bb331e2518c83cf
- [x] **vim-patch:7.4.1840**
  Problem: When using packages an "after" directory cannot be used.
  Solution: Add the "after" directory of the package to 'runtimepath' if it exists
  
  https://github.com/vim/vim/commit/a57024453115592b8847af40ddd965a33898e390
- [x] **vim-patch:7.4.1973**
  Problem:    On MS-Windows the package directory may be added at the end
              because of forward/backward slash differences. (Matthew
              Desjardins)
  Solution:   Ignore slash differences.
  
  https://github.com/vim/vim/commit/4c5717ed8a81f5ae9dfe4f38b17a61fc8421054b
- [x] **vim-patch:7.4.1986**
  Problem:    Compiler warns for loss of data.
  Solution:   Use size_t instead of int. (Christian Brabandt)
  
  https://github.com/vim/vim/commit/fef524bbff9aa186838c35212b2f89f61d627cf8

jamessan committed:
vim-patch:7.4.1384

Problem:    It is not easy to use a set of plugins and their dependencies.
Solution:   Add packages, ":loadopt", 'packpath'.

https://github.com/vim/vim/commit/f6fee0e2d4341c0c2f5339c1268e5877fafd07cf

jamessan committed:
vim-patch:7.4.1388

Problem:    Compiler warning. (Cesar Romani)
Solution:   Initialize variable.

https://github.com/vim/vim/commit/bdcd75275002c3b74015bb9bc0a01b13bb6107d4

jamessan committed:
vim-patch:7.4.1396

Problem:    Compiler warnings for conversions.
Solution:   Add type cast.

https://github.com/vim/vim/commit/1daae446e58fd90f98c51ff3af8f54bfa5197751

jamessan committed:
vim-patch:7.4.1478

Problem:    ":loadplugin" doesn't take care of ftdetect files.
Solution:   Also load ftdetect scripts when appropriate.

https://github.com/vim/vim/commit/1bdd42627d619258d0e847f217cfc1c2795f1ac5

Shougo commented:
More package feature changes are needed.

1479, 1480, 1486, 1492, 1499, 1528, 1550,1552, 1553, 1554, 1596, 1712

jamessan commented:
Yup.  That's why it's `[WIP]`. :) Thanks for the list of patches.  That will save me some time.

Shougo commented:
My work is to list all Vim patches.
All patches are already analyzed :-)

jamessan committed:
vim-patch:7.4.1479

Problem:    No testfor ":loadplugin".
Solution:   Add a test.  Fix how option is being set.

https://github.com/vim/vim/commit/863c1a9079fa340d663ccafb011729a29186d73e

jamessan committed:
vim-patch:7.4.1480

Problem:    Cannot add a pack direcory without loading a plugin.
Solution:   Add the :packadd command.

https://github.com/vim/vim/commit/91715873d19a1859c08eeded7848113596e2f2bd

jamessan committed:
vim-patch:7.4.1486

Problem:    ":loadplugin" is not optimal, some people find it confusing.
Solution:   Only use ":packadd" with an optional "!".

https://github.com/vim/vim/commit/f3654827368e6204608036353a0360e9e7c21e02

jamessan committed:
vim-patch:7.4.1492

Problem:    No command line completion for ":packadd".
Solution:   Implement completion. (Hirohito Higashi)

https://github.com/vim/vim/commit/35ca0e7a1cb6e6daef8e0052a8437801226cef19

jamessan commented:
@justinmk The patches for including the package support include modifications of runtime files related to describing the package functionality.  However, Bram also modified those runtime files (with package-specific content) in some of the intermediate runtime file updates.

For example, the package support was first added in vim/vim@f6fee0e which included changes to doc/repeat.txt.  In vim/vim@328da0d (a runtime update), the file was changed again to include the rename of `:loadplugin` to `:packadd` that had been done in vim/vim@f365482.  This causes a conflict when trying to merge vim/vim@be82c25 because it doesn't have the rename in the runtime files.

How should I handle getting those runtime changes into this PR?

* Cherry-pick just the package-related content from them, leaving the rest of the runtime changes for other merges.
* Pull in all the relevant runtime file updates, cluttering the review of porting the package functionality.

KillTheMule commented:
In my (grantedly) limited experience, runtime patches very much should be merged in order. So if you pull in vim/vim@328da0d, you'd have to pull in all prior ones, which is a real lot, and woud clutter both your schedule and the reviewer's ...

I'm definitely in favor of cherry-picking the package related runtime stuff. If someone else ports the full runtime patch later, it's easy to note the changes are already there. Just try to catch _all_ runtime changes for packages, otherwise you might create a conflict of an earlier runtime patch with your work that isn't as easy to diagnose.

jamessan committed:
vim-patch:f391327

Updated runtime files.

https://github.com/vim/vim/commit/f391327adbbffb11180cf6038a92af1ed144e907

Ignore changes to
* doc/todo.txt: Irrelevant for Neovim
* doc/channel.txt: Channel docs
* doc/tags, syntax/vim.vim: Generated at build time

jamessan committed:
vim-patch:dae8d21

Updated runtime files

https://github.com/vim/vim/commit/dae8d21dd291df6a6679a00be64e18bca0156576

Ignore changes to
* doc/eval.txt: Channel related docs
* doc/help.txt, doc/index.txt, doc/os_390.txt: Removal of obsolete features, which
  already happened in Neovim
* doc/tags: Generated at build time
* doc/todo.txt, doc/version5.txt: Irrelevant to Neovim

jamessan committed:
vim-patch:328da0d

Update runtime files.

https://github.com/vim/vim/commit/328da0dcb7be34b594725eef6dc98d3ea6516d69

Ignore changes to
* doc/channel.txt, doc/eval.txt: Channel related docs
* doc/tags: Generated at build time
* doc/todo.txt: Irrelevant to Neovim

jamessan committed:
vim-patch:7.4.1499

Problem:    No error message when :packadd does not find anything.
Solution:   Add an error message. (Hirohito Higashi)

https://github.com/vim/vim/commit/be82c254862e475a582c0717455e1db6bf96b0d0

jamessan committed:
vim-patch:5f148ec

Update runtime files.

https://github.com/vim/vim/commit/5f148ec0b5a6cedd9129b3abac351034b83cc4f7

Ignore changes to
* doc/channel.txt, doc/eval.txt: Channel related docs
* doc/tags: Generated at build time
* doc/todo.txt: Irrelevant to Neovim

jamessan committed:
vim-patch:7.4.1528

Problem:    Using "ever" for packages is confusing.
Solution:   Use "start", as it's related to startup.

https://github.com/vim/vim/commit/af1a0e371e739f8dff337fd31da0ff8ffb347b43

jamessan committed:
vim-patch:77cdfd1

Updated runtime files.

https://github.com/vim/vim/commit/77cdfd10382e01cc51f4ba1a9177032351843151

Ignore changes to:
* doc/channel.txt, doc/eval.txt: Channel related docs
* doc/options.txt: GUI related docs
* doc/tags: Generated at build time
* doc/todo.txt: Irrelevant for Neovim

jamessan committed:
vim-patch:7.4.1550

Problem:    Cannot load packages early.
Solution:   Add the ":packloadall" command.

https://github.com/vim/vim/commit/2d8f56acb32428d0f965d42dd13b27100b46fa15

jamessan committed:
vim-patch:7.4.1552

Problem:    ":colorscheme" does not use 'packpath'.
Solution:   Also use in "start" and "opt" directories in 'packpath'.

https://github.com/vim/vim/commit/7f8989dd8a627af2185df381195351a913f3777f

jamessan committed:
vim-patch:7.4.1551

Problem:    Cannot generate help tags in all doc directories.
Solution:   Make ":helptags ALL" work.

https://github.com/vim/vim/commit/6bef5306e4f2cacb3a93667992c2312d4b293c9d

jamessan committed:
vim-patch:7.4.1553

Problem:    ":runtime" does not use 'packpath'.
Solution:   Add "what" argument.

https://github.com/vim/vim/commit/8dcf259d904cfb965d31841dc74a5cfaf5a351d9

jamessan committed:
vim-patch:7.4.1554

Problem:    Completion for :colorscheme does not use 'packpath'.
Solution:   Make it work, add a test. (Hirohito Higashi)

https://github.com/vim/vim/commit/52f9c19015df5ee1ee8592b6f3f15b8a57c8f5be

jamessan committed:
vim-patch:7.4.1596

Problem:    Memory leak. (Coverity)
Solution:   Free the pattern.

https://github.com/vim/vim/commit/ba8cd122ef60a7c71a7723be0d635f0c2d4556ab

jamessan committed:
vim-patch:e18c0b3

Updated runtime files.

https://github.com/vim/vim/commit/e18c0b39815c5a746887a509c2cd9f11fadaba07

Ignore changes to
* doc/channel.txt, runtime/tools/demoserver.py: Channel related changes
* doc/if_lua.txt, doc/if_perl.txt, doc/if_pyth.txt, doc/if_ruby.txt,
  doc/if_tcl.txt, doc/todo.txt: Irrelevant to Neovim
* doc/tags: Generated at build time

jamessan committed:
vim-patch:4f3f668

Updated runtime files.

https://github.com/vim/vim/commit/4f3f668c8486444e53163c29d2fc79bf47eb3c82

Ignore changes to
* doc/channel.txt: Channel related docs
* doc/tags: Generated at build time
* doc/todo.txt: Irrelevant to Neovim

jamessan committed:
vim-patch:7.4.1649

Problem:    The matchit plugin needs to be copied to be used.
Solution:   Put the matchit plugin in an optional package.

https://github.com/vim/vim/commit/aedfcbe1e6c7df6edcd6756d7601bfdec7dd2087

Ignore changes to
* Filelist, src/Makefile: Irrelevant to NeoVim
* runtime/vimrc_example.vim, runtime/macros/*, runtime/pack/*: matchit
  is enabled by default in Neovim.

jamessan committed:
Move vimball plugin into an optional package

jamessan committed:
Reword §05.4 to describe using the new vimball package

jamessan committed:
Generate tags and install docs for included packages

oni-link commented on runtime/doc/repeat.txt:
```diff
@@ -373,7 +421,91 @@ Rationale:
+
+Vim will also load ftdetect files, if there are any.
+
+Note that the files under "pack/foo/opt" or not loaded automatically, only the
```
or -> are

jamessan replied:
That's fixed in one of the later patches to be merged.

justinmk commented:
> How should I handle getting those runtime changes into this PR?

There's a third option: put the runtime updates in a separate PR. In general I would recommend merging runtime updates fully, not in pieces, otherwise it makes things complicated.

It's ok if the current runtime at HEAD does not totally match the capabilities of `nvim`. Though we would want to avoid this for a tagged release.

jamessan commented:
At this point, I'm nearly done.  I went with merging both the runtime updates and patches.

Shougo commented:
I think Vim 7.4.1986 is also package patch.

KillTheMule commented:
Maybe soon you need to list the patches you're _not_ including :P

jamessan committed:
vim-patch:7db8f6f

Updated runtime files.

https://github.com/vim/vim/commit/7db8f6f4f85e5d0526d23107b2a5e2334dc23354

Ignore changes to
* doc/tags: Generated at build time
* doc/channel.txt, doc/todo.txt: Irrelevant to Neovim
* pack/dist/opt/matchit/doc/matchit.txt: matchit is enabled by default,
  so description of how to enable isn't needed.

jamessan committed:
vim-patch:7.4.1712

Problem:    For plugins in packages, plugin authors need to take care of all
            dependencies.
Solution:   When loading "start" packages and for :packloadall, first add all
            directories to 'runtimepath' before sourcing plugins.

https://github.com/vim/vim/commit/49b27326447d0827c59c6cd201d58f65c1163086

jamessan committed:
vim-patch:03413f4

Updated runtime files.

https://github.com/vim/vim/commit/03413f44167c4b5cd0012def9bb331e2518c83cf

Ignore changes to
* doc/Makefile, doc/help.txt: Related to Vim's version8 documentation
* doc/gui_x11.txt, doc/todo.txt, doc/vim.1, gvim.desktop, vim.desktop:
  Irrelevant to Neovim
* doc/quickref.txt, doc/options.txt: As of yet unported 'emoji'
* doc/tags, syntax/vim.vim: Generated at build time

jamessan committed:
vim-patch:7.4.1840

Problem:    When using packages an "after" directory cannot be used.
Solution:   Add the "after" directory of the package to 'runtimepath' if it
            exists.

https://github.com/vim/vim/commit/a57024453115592b8847af40ddd965a33898e390

jamessan committed:
vim-patch:7.4.1973

Problem:    On MS-Windows the package directory may be added at the end
            because of forward/backward slash differences. (Matthew
            Desjardins)
Solution:   Ignore slash differences.

https://github.com/vim/vim/commit/4c5717ed8a81f5ae9dfe4f38b17a61fc8421054b

jamessan committed:
vim-patch:7.4.1986

Problem:    Compiler warns for loss of data.
Solution:   Use size_t instead of int. (Christian Brabandt)

https://github.com/vim/vim/commit/fef524bbff9aa186838c35212b2f89f61d627cf8

jamessan committed:
lint

jamessan commented:
I think this is pretty much ready now.

There are more patches like 7.4.1649 where addons from the macros directory are turned into standard packages.  I didn't want to keep all the fun to myself, so I left those out for now.

KillTheMule commented:
> There are more patches like 7.4.1649 where addons from the macros directory are turned into standard packages. I didn't want to keep all the fun to myself, so I left those out for now.

I'd think this would be a nice issue to open and tag entry-level, wouldn't it?

jamessan commented on src/nvim/ex_cmds2.c:
```diff
@@ -2347,11 +2363,12 @@ int do_in_runtimepath(char_u *name, int all, DoInRuntimepathCB callback,
           if (gen_expand_wildcards(1, &buf, &num_files, &files,
-                                   EW_FILE) == OK) {
+                                   (flags & DIP_DIR) ? EW_DIR
+                                                     : EW_FILE) == OK) {
```
I know the linter doesn't like this alignment of the ternary, but I find that far more readable than aligning the `:` with the `(`.
Alternatively, I could assign the value to a variable and use that in the function call.

ZyX-I replied:
If I need to break ternary like this, I break it into three lines, not two. Though in this case it is better to move `== OK` on next line, not part of the ternary.

justinmk commented on src/nvim/ex_getln.c:
```diff
@@ -4212,6 +4221,26 @@ static int ExpandRTDir(char_u *pat, int *num_file, char_u ***file, char *dirname
+    for (int i = 0; dirnames[i] != NULL; i++) {
+      size_t size = STRLEN(dirnames[i]) + pat_len + 20;
+      char_u *s = xmalloc(size);
+      snprintf((char *)s, size, "pack/*/opt/*/%s/%s*.vim", dirnames[i], pat);  // NOLINT
```
As @ZyX-I pointed out in one of my PRs, a way to avoid this NOLINT is:

    "pack/*/opt/""*/%s/%s*.vim"

justinmk replied:
I think in most of the cases here though, it would harm readability to use such a workaround (in some cases multiple times per line).

jamessan replied:
Yeah, I had intended to go back and specify the scope of the NOLINT calls instead of leaving them as plain `NOLINT`.  Forgot about that, though.
